File: dotnet/src/OperationVanguard.Core/Collision/CollisionTypes.cs
using System.Collections.ObjectModel;

namespace OperationVanguard.Core;

/// <summary>Bit mask selecting which categories of geometry a collision query considers.</summary>
[Flags]
public enum CollisionLayer
{
    None = 0,
    /// <summary>Static level geometry.</summary>
    World = 1 << 0,
    /// <summary>Brushes that stop players but let bullets through.</summary>
    PlayerClip = 1 << 1,
    /// <summary>Brushes that stop bullets but not players.</summary>
    BulletClip = 1 << 2,
    /// <summary>Player capsules.</summary>
    Player = 1 << 3,
    /// <summary>Deployables, care packages, and sentry guns.</summary>
    Deployable = 1 << 4,
    /// <summary>Killstreak vehicles.</summary>
    Vehicle = 1 << 5,
    /// <summary>Destructible props.</summary>
    Breakable = 1 << 6,
    /// <summary>Non-solid water volumes that affect movement and audio.</summary>
    Water = 1 << 7,

    /// <summary>Everything a walking player collides with.</summary>
    Movement = World | PlayerClip | Breakable | Deployable | Vehicle,
    /// <summary>Everything a bullet can hit.</summary>
    Bullet = World | BulletClip | Player | Breakable | Deployable | Vehicle,
    /// <summary>Everything that blocks line of sight.</summary>
    Sight = World | Breakable,
    /// <summary>Everything a thrown grenade bounces off.</summary>
    Projectile = World | PlayerClip | Breakable | Deployable | Vehicle,
}

/// <summary>The result of casting a ray through the collision world.</summary>
public sealed class RaycastHit
{
    /// <summary>True if anything was hit. Check this before reading the remaining fields.</summary>
    public bool Hit { get; set; }

    /// <summary>Distance along the ray to the impact point, in metres.</summary>
    public double Distance { get; set; }

    public Vec3 Point { get; set; } = new();

    public Vec3 Normal { get; set; } = new(0d, 1d, 0d);

    public SurfaceType Surface { get; set; } = SurfaceType.Concrete;

    /// <summary>Entity id for a dynamic hit, or zero for static geometry.</summary>
    public int Entity { get; set; }

    /// <summary>Index of the brush that was hit, or -1 for entities.</summary>
    public int BrushIndex { get; set; } = -1;

    /// <summary>Material thickness the ray must cross to continue, in metres.</summary>
    public double Thickness { get; set; }

    public CollisionLayer Layer { get; set; } = CollisionLayer.None;
}

/// <summary>The result of sweeping a vertical capsule through the collision world.</summary>
public sealed class SweepHit
{
    public bool Hit { get; set; }

    /// <summary>Fraction of the requested motion completed before contact, in [0, 1].</summary>
    public double Fraction { get; set; } = 1d;

    public Vec3 Point { get; set; } = new();

    public Vec3 Normal { get; set; } = new(0d, 1d, 0d);

    public SurfaceType Surface { get; set; } = SurfaceType.Concrete;

    public int Entity { get; set; }

    public int BrushIndex { get; set; } = -1;

    /// <summary>True when the capsule started already overlapping geometry.</summary>
    public bool StartedSolid { get; set; }
}

/// <summary>Optional entity filtering applied in addition to a collision layer mask.</summary>
public sealed class QueryFilter
{
    public QueryFilter()
    {
    }

    public QueryFilter(CollisionLayer layers)
    {
        Layers = layers;
    }

    public CollisionLayer Layers { get; set; }

    /// <summary>Entity ids to skip before evaluating <see cref="EntityPredicate"/>.</summary>
    public IReadOnlyCollection<int>? IgnoreEntities { get; set; }

    /// <summary>Return false to skip a candidate entity.</summary>
    public Func<int, bool>? EntityPredicate { get; set; }

    /// <summary>
    /// Returns whether a candidate entity passes this filter. Ignored ids short-circuit
    /// before the predicate, matching the TypeScript collision backend.
    /// </summary>
    public bool IncludesEntity(int id)
    {
        if (IgnoreEntities is not null && IgnoreEntities.Contains(id))
        {
            return false;
        }

        return EntityPredicate?.Invoke(id) is not false;
    }
}

/// <summary>
/// Deterministic collision queries shared by movement, ballistics, AI, spawning,
/// and projectile physics. Directions are unit length and distances are metres.
/// </summary>
public interface ICollisionWorld
{
    /// <summary>Fills and returns <paramref name="output"/> with the nearest ray hit.</summary>
    RaycastHit Raycast(
        Vec3 origin,
        Vec3 direction,
        double maxDistance,
        QueryFilter filter,
        RaycastHit output);

    /// <summary>
    /// Writes ray hits nearest-first and returns the number written, up to both
    /// <paramref name="maxHits"/> and the capacity of <paramref name="output"/>.
    /// </summary>
    int RaycastAll(
        Vec3 origin,
        Vec3 direction,
        double maxDistance,
        QueryFilter filter,
        IList<RaycastHit> output,
        int maxHits);

    /// <summary>
    /// Sweeps a vertical capsule, whose feet begin at <paramref name="start"/>,
    /// along <paramref name="delta"/> and returns the first contact.
    /// </summary>
    SweepHit SweepCapsule(
        Vec3 start,
        double height,
        double radius,
        Vec3 delta,
        QueryFilter filter,
        SweepHit output);

    /// <summary>
    /// Pushes a capsule out of overlapping geometry, writes the corrected feet
    /// position to <paramref name="output"/>, and returns true if it moved.
    /// </summary>
    bool ResolvePenetration(
        Vec3 position,
        double height,
        double radius,
        QueryFilter filter,
        Vec3 output);

    /// <summary>Cheap capsule occupancy test used by spawn selection.</summary>
    bool IsCapsuleFree(Vec3 position, double height, double radius, QueryFilter filter);

    /// <summary>Returns whether the line segment between two points is unobstructed.</summary>
    bool IsVisible(Vec3 from, Vec3 to, QueryFilter filter);

    /// <summary>Ground height below a point, or negative infinity if no ground is found.</summary>
    double GroundHeightAt(double x, double z, double fromY, double maxDrop);
}

/// <summary>Collision result factories and the canonical per-surface behaviour/render tables.</summary>
public static class CollisionTypes
{
    public static RaycastHit CreateRaycastHit() => new();

    public static SweepHit CreateSweepHit() => new();

    public static IReadOnlyDictionary<SurfaceType, SurfaceProperties> SurfaceProperties { get; } =
        ReadOnly(new Dictionary<SurfaceType, SurfaceProperties>
        {
            [SurfaceType.Concrete] = Properties(0.25d, 0.35d, 0.85d, false),
            [SurfaceType.Metal] = Properties(0.15d, 0.2d, 1d, false),
            [SurfaceType.Wood] = Properties(0.7d, 0.75d, 0.8d, true),
            [SurfaceType.Dirt] = Properties(0.4d, 0.45d, 0.55d, false),
            [SurfaceType.Grass] = Properties(0.5d, 0.55d, 0.4d, false),
            [SurfaceType.Sand] = Properties(0.35d, 0.4d, 0.45d, false),
            [SurfaceType.Water] = Properties(0.9d, 0.6d, 0.9d, false),
            [SurfaceType.Glass] = Properties(0.95d, 0.92d, 0.9d, true),
            [SurfaceType.Foliage] = Properties(1d, 0.98d, 0.6d, false),
            [SurfaceType.Flesh] = Properties(0.8d, 0.7d, 0.3d, false),
            [SurfaceType.Carpet] = Properties(0.6d, 0.7d, 0.3d, false),
            [SurfaceType.Gravel] = Properties(0.3d, 0.4d, 1d, false),
            [SurfaceType.Snow] = Properties(0.6d, 0.6d, 0.5d, false),
            [SurfaceType.Tile] = Properties(0.45d, 0.5d, 0.95d, true),
            [SurfaceType.Plastic] = Properties(0.8d, 0.85d, 0.7d, true),
            [SurfaceType.Brick] = Properties(0.3d, 0.4d, 0.85d, false),
        });

    /// <summary>Default 24-bit RGB albedo when a brush does not override it.</summary>
    public static IReadOnlyDictionary<SurfaceType, int> SurfaceColors { get; } =
        ReadOnly(new Dictionary<SurfaceType, int>
        {
            [SurfaceType.Concrete] = 0x8a8a86,
            [SurfaceType.Metal] = 0x6e747a,
            [SurfaceType.Wood] = 0x8a6440,
            [SurfaceType.Dirt] = 0x6b5a45,
            [SurfaceType.Grass] = 0x5c7a44,
            [SurfaceType.Sand] = 0xc2ab7f,
            [SurfaceType.Water] = 0x2b5a72,
            [SurfaceType.Glass] = 0xa8c4cc,
            [SurfaceType.Foliage] = 0x40632f,
            [SurfaceType.Flesh] = 0x9c5a52,
            [SurfaceType.Carpet] = 0x6a4a48,
            [SurfaceType.Gravel] = 0x7d7a72,
            [SurfaceType.Snow] = 0xe4eaf0,
            [SurfaceType.Tile] = 0xb8b4ac,
            [SurfaceType.Plastic] = 0x585c62,
            [SurfaceType.Brick] = 0x91564a,
        });

    /// <summary>Default physical-shader roughness by surface.</summary>
    public static IReadOnlyDictionary<SurfaceType, double> SurfaceRoughness { get; } =
        ReadOnly(new Dictionary<SurfaceType, double>
        {
            [SurfaceType.Concrete] = 0.92d,
            [SurfaceType.Metal] = 0.42d,
            [SurfaceType.Wood] = 0.78d,
            [SurfaceType.Dirt] = 0.98d,
            [SurfaceType.Grass] = 0.95d,
            [SurfaceType.Sand] = 0.96d,
            [SurfaceType.Water] = 0.08d,
            [SurfaceType.Glass] = 0.05d,
            [SurfaceType.Foliage] = 0.88d,
            [SurfaceType.Flesh] = 0.7d,
            [SurfaceType.Carpet] = 0.99d,
            [SurfaceType.Gravel] = 0.96d,
            [SurfaceType.Snow] = 0.85d,
            [SurfaceType.Tile] = 0.35d,
            [SurfaceType.Plastic] = 0.55d,
            [SurfaceType.Brick] = 0.9d,
        });

    /// <summary>
    /// Default physical-shader metalness by surface. Metal deliberately remains
    /// at 0.3 because the asset-free renderer has no environment map to reflect.
    /// </summary>
    public static IReadOnlyDictionary<SurfaceType, double> SurfaceMetalness { get; } =
        ReadOnly(new Dictionary<SurfaceType, double>
        {
            [SurfaceType.Concrete] = 0d,
            [SurfaceType.Metal] = 0.3d,
            [SurfaceType.Wood] = 0d,
            [SurfaceType.Dirt] = 0d,
            [SurfaceType.Grass] = 0d,
            [SurfaceType.Sand] = 0d,
            [SurfaceType.Water] = 0.1d,
            [SurfaceType.Glass] = 0d,
            [SurfaceType.Foliage] = 0d,
            [SurfaceType.Flesh] = 0d,
            [SurfaceType.Carpet] = 0d,
            [SurfaceType.Gravel] = 0d,
            [SurfaceType.Snow] = 0d,
            [SurfaceType.Tile] = 0.05d,
            [SurfaceType.Plastic] = 0d,
            [SurfaceType.Brick] = 0d,
        });

    private static SurfaceProperties Properties(
        double penetration,
        double damageRetention,
        double footstepVolume,
        bool breakable) =>
        new()
        {
            Penetration = penetration,
            DamageRetention = damageRetention,
            FootstepVolume = footstepVolume,
            Breakable = breakable,
        };

    private static IReadOnlyDictionary<TKey, TValue> ReadOnly<TKey, TValue>(
        Dictionary<TKey, TValue> values)
        where TKey : notnull => new ReadOnlyDictionary<TKey, TValue>(values);
}

File: web/src/shared/collision/collision-types.ts
/**
 * Collision query interface.
 *
 * The movement controller, ballistics, AI line-of-sight and grenade physics all
 * talk to the world through this interface and nothing else. That keeps the
 * broadphase implementation swappable and — critically — keeps the simulation
 * free of any three.js dependency so it can run on the server.
 *
 * Convention: all directions passed in are unit length; all distances are metres.
 */

import type { Vec3 } from '../math.js';
import type { EntityId, SurfaceType } from '../types.js';

/** Bit mask selecting which categories of geometry a query should consider. */
export enum CollisionLayer {
  None = 0,
  /** Static level geometry. */
  World = 1 << 0,
  /** Brushes that stop players but let bullets through (railings, chain fence). */
  PlayerClip = 1 << 1,
  /** Brushes that stop bullets but not players (invisible bullet blockers). */
  BulletClip = 1 << 2,
  /** Player capsules. */
  Player = 1 << 3,
  /** Deployables, care packages, sentry guns. */
  Deployable = 1 << 4,
  /** Killstreak vehicles. */
  Vehicle = 1 << 5,
  /** Destructible props. */
  Breakable = 1 << 6,
  /** Water volumes — no collision, but changes movement and audio. */
  Water = 1 << 7,

  /** Everything a walking player collides with. */
  Movement = World | PlayerClip | Breakable | Deployable | Vehicle,
  /** Everything a bullet can hit. */
  Bullet = World | BulletClip | Player | Breakable | Deployable | Vehicle,
  /** Everything that blocks line of sight. */
  Sight = World | Breakable,
  /** Everything a thrown grenade bounces off. */
  Projectile = World | PlayerClip | Breakable | Deployable | Vehicle,
}

export interface RaycastHit {
  /** True if anything was hit. Check this before reading any other field. */
  hit: boolean;
  /** Distance along the ray to the impact point. */
  distance: number;
  point: Vec3;
  normal: Vec3;
  surface: SurfaceType;
  /** Set when the hit was against an entity rather than static geometry. */
  entity: EntityId;
  /** Index of the brush that was hit, or -1 for entities. */
  brushIndex: number;
  /** How much material the ray would have to pass through to continue, in metres. */
  thickness: number;
  /** Which layer the hit belongs to. */
  layer: CollisionLayer;
}

export function createRaycastHit(): RaycastHit {
  return {
    hit: false,
    distance: 0,
    point: { x: 0, y: 0, z: 0 },
    normal: { x: 0, y: 1, z: 0 },
    surface: 0 as SurfaceType,
    entity: 0,
    brushIndex: -1,
    thickness: 0,
    layer: CollisionLayer.None,
  };
}

export interface SweepHit {
  hit: boolean;
  /** Fraction of the requested motion completed before contact, in [0, 1]. */
  fraction: number;
  point: Vec3;
  normal: Vec3;
  surface: SurfaceType;
  entity: EntityId;
  brushIndex: number;
  /** True when the capsule started already overlapping geometry. */
  startedSolid: boolean;
}

export function createSweepHit(): SweepHit {
  return {
    hit: false,
    fraction: 1,
    point: { x: 0, y: 0, z: 0 },
    normal: { x: 0, y: 1, z: 0 },
    surface: 0 as SurfaceType,
    entity: 0,
    brushIndex: -1,
    startedSolid: false,
  };
}

/** Optional filter so a query can ignore the shooter's own capsule, teammates, etc. */
export interface QueryFilter {
  layers: CollisionLayer;
  /** Entity ids to skip. */
  ignoreEntities?: readonly EntityId[];
  /** Return false to skip a candidate entity. */
  entityPredicate?: (id: EntityId) => boolean;
}

/**
 * The queries every collision backend must provide.
 *
 * Implementations must be deterministic: given the same world state and the same
 * arguments they must return bit-identical results on client and server, because
 * prediction replays them.
 */
export interface CollisionWorld {
  /**
   * Cast a ray and return the nearest hit.
   * `out` is filled in and returned so hot loops don't allocate.
   */
  raycast(
    origin: Vec3,
    direction: Vec3,
    maxDistance: number,
    filter: QueryFilter,
    out: RaycastHit,
  ): RaycastHit;

  /**
   * Cast a ray and report every hit along it, nearest first. Used for bullet
   * penetration, where a round may pass through several surfaces.
   * Returns the number of hits written into `out`.
   */
  raycastAll(
    origin: Vec3,
    direction: Vec3,
    maxDistance: number,
    filter: QueryFilter,
    out: RaycastHit[],
    maxHits: number,
  ): number;

  /**
   * Sweep a vertical capsule (feet at `start`, given height and radius) along
   * `delta` and report first contact.
   */
  sweepCapsule(
    start: Vec3,
    height: number,
    radius: number,
    delta: Vec3,
    filter: QueryFilter,
    out: SweepHit,
  ): SweepHit;

  /**
   * Push a capsule out of any geometry it currently overlaps.
   * Writes the corrected feet position into `out` and returns true if it moved.
   */
  resolvePenetration(
    position: Vec3,
    height: number,
    radius: number,
    filter: QueryFilter,
    out: Vec3,
  ): boolean;

  /** Cheap "is this capsule position free?" test used by spawn selection. */
  isCapsuleFree(position: Vec3, height: number, radius: number, filter: QueryFilter): boolean;

  /**
   * Unobstructed line of sight between two points. Separate from `raycast`
   * because AI calls it thousands of times per second and it can short-circuit.
   */
  isVisible(from: Vec3, to: Vec3, filter: QueryFilter): boolean;

  /** Ground height directly beneath a point, or -Infinity if there is none. */
  groundHeightAt(x: number, z: number, fromY: number, maxDrop: number): number;
}

// ---------------------------------------------------------------------------
// Surface behaviour table
//
// Lives here rather than in types.ts because it is consumed almost exclusively
// by collision-adjacent code: penetration, footsteps, impact VFX.
// ---------------------------------------------------------------------------

import type { SurfaceProperties } from '../types.js';
import { SurfaceType as ST } from '../types.js';

export const SURFACE_PROPERTIES: Record<SurfaceType, SurfaceProperties> = {
  [ST.Concrete]: { penetration: 0.25, damageRetention: 0.35, footstepVolume: 0.85, breakable: false },
  [ST.Metal]: { penetration: 0.15, damageRetention: 0.2, footstepVolume: 1.0, breakable: false },
  [ST.Wood]: { penetration: 0.7, damageRetention: 0.75, footstepVolume: 0.8, breakable: true },
  [ST.Dirt]: { penetration: 0.4, damageRetention: 0.45, footstepVolume: 0.55, breakable: false },
  [ST.Grass]: { penetration: 0.5, damageRetention: 0.55, footstepVolume: 0.4, breakable: false },
  [ST.Sand]: { penetration: 0.35, damageRetention: 0.4, footstepVolume: 0.45, breakable: false },
  [ST.Water]: { penetration: 0.9, damageRetention: 0.6, footstepVolume: 0.9, breakable: false },
  [ST.Glass]: { penetration: 0.95, damageRetention: 0.92, footstepVolume: 0.9, breakable: true },
  [ST.Foliage]: { penetration: 1.0, damageRetention: 0.98, footstepVolume: 0.6, breakable: false },
  [ST.Flesh]: { penetration: 0.8, damageRetention: 0.7, footstepVolume: 0.3, breakable: false },
  [ST.Carpet]: { penetration: 0.6, damageRetention: 0.7, footstepVolume: 0.3, breakable: false },
  [ST.Gravel]: { penetration: 0.3, damageRetention: 0.4, footstepVolume: 1.0, breakable: false },
  [ST.Snow]: { penetration: 0.6, damageRetention: 0.6, footstepVolume: 0.5, breakable: false },
  [ST.Tile]: { penetration: 0.45, damageRetention: 0.5, footstepVolume: 0.95, breakable: true },
  [ST.Plastic]: { penetration: 0.8, damageRetention: 0.85, footstepVolume: 0.7, breakable: true },
  [ST.Brick]: { penetration: 0.3, damageRetention: 0.4, footstepVolume: 0.85, breakable: false },
};

/** Default albedo per surface, used when a brush doesn't override it. */
export const SURFACE_COLORS: Record<SurfaceType, number> = {
  [ST.Concrete]: 0x8a8a86,
  [ST.Metal]: 0x6e747a,
  [ST.Wood]: 0x8a6440,
  [ST.Dirt]: 0x6b5a45,
  [ST.Grass]: 0x5c7a44,
  [ST.Sand]: 0xc2ab7f,
  [ST.Water]: 0x2b5a72,
  [ST.Glass]: 0xa8c4cc,
  [ST.Foliage]: 0x40632f,
  [ST.Flesh]: 0x9c5a52,
  [ST.Carpet]: 0x6a4a48,
  [ST.Gravel]: 0x7d7a72,
  [ST.Snow]: 0xe4eaf0,
  [ST.Tile]: 0xb8b4ac,
  [ST.Plastic]: 0x585c62,
  [ST.Brick]: 0x91564a,
};

/** Default PBR roughness per surface. */
export const SURFACE_ROUGHNESS: Record<SurfaceType, number> = {
  [ST.Concrete]: 0.92,
  [ST.Metal]: 0.42,
  [ST.Wood]: 0.78,
  [ST.Dirt]: 0.98,
  [ST.Grass]: 0.95,
  [ST.Sand]: 0.96,
  [ST.Water]: 0.08,
  [ST.Glass]: 0.05,
  [ST.Foliage]: 0.88,
  [ST.Flesh]: 0.7,
  [ST.Carpet]: 0.99,
  [ST.Gravel]: 0.96,
  [ST.Snow]: 0.85,
  [ST.Tile]: 0.35,
  [ST.Plastic]: 0.55,
  [ST.Brick]: 0.9,
};

/**
 * Metalness per surface, for the physical shader.
 *
 * `Metal` is deliberately far below the ~0.9 a real metal would take. A metal
 * has no diffuse response: under a physical shader it is only what it reflects,
 * and this game ships zero binary assets and therefore has no environment map.
 * At a physically honest value every metal surface the sun does not directly
 * strike renders pure black — a shaded steel warehouse becomes a hole cut in the
 * world, and an underground station's rolling stock becomes a silhouette. A
 * painted, weathered value is the correct trade for a game with nothing to
 * reflect.
 */
export const SURFACE_METALNESS: Record<SurfaceType, number> = {
  [ST.Concrete]: 0,
  [ST.Metal]: 0.3,
  [ST.Wood]: 0,
  [ST.Dirt]: 0,
  [ST.Grass]: 0,
  [ST.Sand]: 0,
  [ST.Water]: 0.1,
  [ST.Glass]: 0,
  [ST.Foliage]: 0,
  [ST.Flesh]: 0,
  [ST.Carpet]: 0,
  [ST.Gravel]: 0,
  [ST.Snow]: 0,
  [ST.Tile]: 0.05,
  [ST.Plastic]: 0,
  [ST.Brick]: 0,
};
